# Исследование книжного сервиса

Hаша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


Нам дано:  
Таблица books, содержит данные о книгах

Таблица authors, содержит данные об авторах

Таблица publishers, содержит данные об издательствах

Таблица ratings, содержит данные о пользовательских оценках книг

Таблица reviews, содержит данные о пользовательских обзорах на книги

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [2]:
books = '''SELECT * 
FROM books'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(books), con = con)
# откроем и просмотрим таблицу "books"

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [3]:
ratings = '''SELECT * 
FROM ratings'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(ratings), con = con)
# откроем и просмотрим таблицу "rating"

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [4]:
authors = '''SELECT * 
FROM authors'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(authors), con = con)
# откроем и просмотрим таблицу "authors"

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [5]:
reviews = '''SELECT * 
FROM reviews'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(reviews), con = con)
# откроем и просмотрим таблицу "reviews"

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [6]:
publishers = '''SELECT * 
FROM publishers'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(publishers), con = con)
# откроем и просмотрим таблицу "publishers"

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [7]:
books_01012000 = '''
SELECT COUNT (book_id) 
FROM books
WHERE CAST(publication_date AS date) >'2000-01-01'

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(books_01012000), con = con)

count
0    819

После 01 января 2000 года (не считая само 01.01.2000) вышло 819 книг

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [8]:
avg_obzor = '''
SELECT b.title,
       b.book_id,
       ROUND(AVG(rt.rating), 2) AS avg_rating,
       COUNT(DISTINCT rw.review_id) AS cnt_of_reviews

FROM books AS b

LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
LEFT JOIN reviews AS rw ON b.book_id = rw.book_id

GROUP BY b.title, b.book_id
ORDER BY cnt_of_reviews DESC
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(avg_obzor), con = con)

title  book_id  avg_rating  \
0                              Twilight (Twilight  #1)      948        3.66   
1                                  Water for Elephants      963        3.98   
2                                     The Glass Castle      734        4.21   
3    Harry Potter and the Prisoner of Azkaban (Harr...      302        4.41   
4    The Curious Incident of the Dog in the Night-Time      695        4.08   
..                                                 ...      ...         ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel       83        3.67   
996                            The Natural Way to Draw      808        3.00   
997   The Cat in the Hat and Other Dr. Seuss Favorites      672        5.00   
998                          Essential Tales and Poems      221        4.00   
999  Disney's Beauty and the Beast (A Little Golden...      191        4.00   

     cnt_of_reviews  
0                 7  
1                 6  
2                 6  
3                 6  
4                 6  
..              ...  
995               0  
996               0  
997               0  
998               0  
999               0  

[1000 rows x 4 columns]

Средние оценки книг лежат в диапазоне от 5.0 до 1.5, количество обзоров от 7 до 0

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [9]:
pub = '''
SELECT p.publisher,
       COUNT(b.book_id) AS total_books
       
FROM books AS b

LEFT JOIN  publishers AS p ON b.publisher_id = p.publisher_id

WHERE num_pages > 50      

GROUP BY p.publisher
ORDER BY total_books DESC
LIMIT 1

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(pub), con = con)

publisher  total_books
0  Penguin Books           42

Penguin Books - это издательство, которое выпустило наибольшее число книг толще 50 страниц

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [10]:
top_author = '''
SELECT 
    aa.author,
    AVG(aa.avg_rating) AS t_rating


FROM  
    (SELECT a.author AS author,
            AVG(r.rating) AS avg_rating
        
        FROM books AS b
        
            LEFT JOIN authors AS a ON b.author_id = a.author_id
            LEFT JOIN ratings AS r ON b.book_id = r.book_id
       
       GROUP BY  b.book_id, a.author_id
             
        HAVING  COUNT(r.rating_id) >=50 
        ) AS aa 

GROUP BY        aa.author
ORDER BY     t_rating DESC
LIMIT 1

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(top_author), con = con)

author  t_rating
0  J.K. Rowling/Mary GrandPré  4.283844

 J.K. Rowling/Mary GrandPré	- автор с самой высокой средней оценкой книг 4.28 

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [11]:
top_users = '''
SELECT AVG(r.review_cnt) AS avg_revie

FROM(SELECT COUNT(review_id) AS review_cnt,
                 username

      FROM reviews

      WHERE username IN  (SELECT username
                     FROM ratings
                     GROUP BY username
                     HAVING COUNT(ratings)> 48)
      GROUP BY username) AS r





'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(top_users), con = con)

avg_revie
0       24.0

24 - это среднее количество обзоров от пользователей, которые поставили больше 48 оценок

### Вывод:
В этой работе мы проанализировали 5 таблиц с данными о книгах, авторах, издательствах, оценках на книги и обзорах.
После всего мы можем сделать вывод, что после 01 января 2000 года  вышло 819 книг
Средние оценки книг лежат в диапазоне от 5.0 до 1.5, количество обзоров от 7 до 0
Penguin Books - это издательство, которое выпустило наибольшее число книг толще 50 страниц
J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой книг 4.28
24 - это среднее количество обзоров от пользователей, которые поставили больше 48 оценок